[FLINK-11726][network] Refactor the creation of ResultPartition and InputGate into NetworkEnvironment

zhijiangW opened the pull request:
## What is the purpose of the change

*The `ResultPartition` and `InputGate` are created in `Task` currently. Based on pluggable `ShuffleManager`, they should be created via `ShuffleService`.*

*The `NetworkEnvironment` would be refactored into `NetworkShuffleService` future. So we could migrate the process of creating `ResultPartition` and `InputGate` into current `NetworkEnvironment`. The `BufferPool` and metrics setup can also be done along with the creations. Then we can remove the process of `NetworkEnvironment#registerTask`.*

## Brief change log

 - *Pass `IOManager` and `enableDetailedNetworkMetrics` in the constructor of `NetworkEnvironment`*
  - *Introduce `TaskId` field for identifying the unique task in `TaskInfo`*
  - *Introduce `NetworkEnvironment#createResultPartitionWriter` and `NetworkEnvironment#createInputGate` to refactor the process of creation of `ResultPartition` and `SingleInputGate` in `Task`*
  - *Modify the `TaskIOMetricGroup` for setting up the input and output metrics*
  - *Remove the explicit reference with `ResultPartition` and `SingleInputGate` in `Task`, which can be got directly from `NetworkEnvironment` by `TaskId`*

## Verifying this change

This change is already covered by existing tests, such as *NetworkEnvironmentTest*, *TaskTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -106,28 +128,32 @@ public NetworkEnvironment(
+			int partitionRequestMaxBackoff,
+			int networkBuffersPerChannel,
+			int extraNetworkBuffersPerGate,
+			boolean enableNetworkDetailedMetrics,
```
Another option might we could pass `NetworkConfiguration` into the constructor instead of these specific parameters `partitionRequestInitialBackoff`, `partitionRequestMaxBackoff`, `networkBuffersPerChannel`, `extraNetworkBuffersPerGate`, `enableNetworkDetailedMetrics`, `enableCreditBased`. Then it seems better to not change this constructor always to add new parameters, just extend the `NetworkConfiguration`.

azagrebin replied:
I think it is good idea. At the end, `ShuffleManager.createShuffleService(config)` will probably need to accept Flink config to allow users set specific options for certain shuffle service.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -202,28 +233,97 @@ public TaskKvStateRegistry createKvStateTaskRegistry(JobID jobId, JobVertexID jo
+	public ResultPartitionWriter[] createResultPartitionWriters(
+			String taskId,
+			JobID jobId,
+			ExecutionAttemptID executionId,
```
I think we could refactor the `ResultPartitionDeploymentDescriptor` to contain `ResultPartitionID` directly, then it might seem better to remove `ExecutionAttemptID` here as a final form of `ShuffleService#createResultPartitionWriter`.  I plan to do it in a separate issue.

zhijiangW replied:
This issue is done by #7835

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -202,28 +233,97 @@ public TaskKvStateRegistry createKvStateTaskRegistry(JobID jobId, JobVertexID jo
+		return allPartitions.get(taskId);
+	}
+
+	public SingleInputGate[] getInputGates(String taskId) {
```
The returned type could be changed to `InputGate[]` and `ResultPartitionWriter[]` after #7767 , #7785  merged.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -88,6 +104,12 @@
 
+	private final boolean enableNetworkDetailedMetrics;
+
+	private final ConcurrentHashMap<String, ResultPartition[]> allPartitions;
```
resultPartitionManager already has "partition by execution id" mapping.
why do we need to introduce extra `taskId`? task seems to be identifiable by execution id.

zhijiangW replied:
Actually there is another consideration for this issue. 
Currently we pass both `taskId` and `executionId` in the method of `ShuffleService#createResultPartitionWriters`. The parameters seem a bit duplicated  and not clean. The `taskId` is used internally in `ResultPartition` for debugging log and it is more readable than `executionId`. The `executionId` parameter could be removed future from this method if it is contained directly from `ResultPartitionDeploymentDescriptor` which is proposed in #7835 . So I used `taskId` for index here for that consideration.
If we do not need `taskId` for debugging info in `ResultPartition`, then we can remove this parameter and only retain `executionId` for index. Or we retain both parameters as now here and change to use `executionId` for index, then in future we could consider how to handle `taskId` to make the interface simple?

azagrebin replied:
At the first glance, it looks like we could include task name into `ResultPartitionDeploymentDescriptor` as well in #7835 along with `executionId`.
I would still index by `executionId` if needed (add to method at the moment before #7835), because it is more reliable and concise than task name + id used for debug.

zhijiangW replied:
Sorry for missing this comment before.

Yes, I agree with the point of indexing by `executionId`, and use task name + id for debugging.
My current thought is passing `ShuffleTaskInfo` into `ShuffleService` during creating partition/gate, so #7835 would not be used any more. I would double check the details from the implementation.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -202,28 +233,97 @@ public TaskKvStateRegistry createKvStateTaskRegistry(JobID jobId, JobVertexID jo
 		return kvStateRegistry.createTaskRegistry(jobId, jobVertexId);
 	}
 
+	public ResultPartition[] getResultPartitionWriters(String taskId) {
```
will the ShuffleService also have the partition/gate getters then? it seems logical to let ShuffleService manage partitions/gates. Alternatively, task could still keep final arrays with partitions/gates. Then there will be less dependencies on ShuffleService interface. What do you think?

zhijiangW replied:
That is a good idea for removing gutters from `ShuffleService` to make interface simple, then let task still maintain the arrays.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -202,28 +233,97 @@ public TaskKvStateRegistry createKvStateTaskRegistry(JobID jobId, JobVertexID jo
-				setupPartition(partition);
+		allPartitions.put(taskId, parititons);
+
+		metrics.initializeOutputBufferMetrics(parititons);
```
Initially, we thought that ShuffleService would create separately partitions/gates.
Now it seems that metrics and partition manager keep partitions/gates by task.
It might change then ShuffleService interface.

As one of steps, we should probably refactor out netty specifics from `TaskIOMetricGroup` and register those specific metrics inside netty ShuffleService.

zhijiangW replied:
Yes, this change is not considered before. The current `IOMetrics` relies on specific `ResultPartition` not regular `ResultPartitionWriter` interface. The final form of ShuffleService interface might include `{taskIdentifier(executionId or taskId), RPDD, metrics}`.

It would seem better if we could make `IOMetricGroup` created inside netty `ShuffleService` to make interface clean. But I have not thought through whether it is feasible to do like that.
From another aspect, if we pass `IOMetricGroup` to the interface as now, it might also seem reasonable because the regular metric framework could be also suitable for other writers. I referred to `ShuffleManager#getReader|getWriter` in spark before, and it also has the similar metric component in the interface.

After the metrics initialization is migrated into `ShuffleService` stack as now, `Task` can only reference with regular `ResultPartitionWriter` and `InputGate` arrays to solve some issues in #7549 .

Do you have further suggestions for metrics?

azagrebin replied:
Actually. looking more into `ResultPartitionManager`, we use indexing by execution id eventually only from task and task already knows its partitions/gates. Maybe, if we move all these partition/gate release/close for-loops from `NetworkEnvironment.unregisterTask` and `ResultPartitionManager.releasePartitionsProducedBy` to `Task`, this would help further to simplify future ShuffleService. If needed, netty `ResultPartition` could remove itself from `ResultPartitionManager` in `release()` method. This would be one step forward to get rid of "by execution id" indexing in `NetworkEnvironment`/`ResultPartitionManager`.

Then we have only metrics with "task to partitions/gates" dependency. Here, it does not seem to be trivial. In general, looking into `TaskIOMetricGroup.initializeOutputBufferMetrics`, `createResultPartitionWriters/gates` could just accept `ProxyMetricGroup<?>` instead of `TaskIOMetricGroup`. Then `initializeOutputBufferMetrics` could go into `NetworkEnvironment` using `ProxyMetricGroup.addGroup` with all netty specific gauge classes.

Thinking more about register "per task" VS "per partition/gate", "per task" is actually more restrictive and we do not need Shuffle API to be more flexible for Task. Other shuffle services might also benefit from this way of registration.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -292,40 +392,31 @@ public void unregisterTask(Task task) {
 			}
 
-			for (ResultPartition partition : task.getProducedPartitions()) {
+			for (ResultPartition partition : allPartitions.get(task.getTaskInfo().getTaskId())) {
```
If we had just execution id to index task partitions/gates, `unregisterTask` could just accept `ExecutionId` and cause in case of failure. The first log statement with task specifics could go outside, maybe. Just less dependency on Task class.

zhijiangW replied:
Yes, I agree with your idea. 
Even though we remove the first log and use `executionId` for index, we still need to consider how to refactor the uses of `task.isCanceledOrFailed()` and `task.getFailureCause()` in this method. I would try to refactor it in this way.

azagrebin replied:
the method could also accept `cause` additionally to id. `cause != null` would mean failure case.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -421,6 +509,25 @@ public void shutdown() {
 
+			for (ResultPartition[] resultPartitions : allPartitions.values()) {
+				for (ResultPartition partition : resultPartitions) {
+					partition.destroyBufferPool();
```
Looks like, these partition/gates releases were supposed to happen on task cancel or finish.
Do you think we should have them again here? to double-check everything is released? :)

zhijiangW replied:
This is a good question. 

In current implementation, if the task is finished then `NetworkEnvironment#unregisterTask` is called to release resources of partitions/gates. So this part should be retained in the case of finished task.

If the task is canceled then the partitions/gates are released directly in canceler thread and `unregisterTask` would be called finally before task exits. For `unregisterTask`, it only works for `TaskEventDispatcher` and `ResultPartitionManager`.

In future the partition/gate release should be only managed by `ShuffleService`, not exposed to outside. The canceler thread could also call `ShuffleService#unregisterTask` to release from all the components including `TaskEventDispatcher` and `ResultPartitionManager`. Then it seems reasonable to call `ShuffleService#unregisterTask` twice in task thread and canceler thread separately. What do you think or have better options?

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -299,7 +296,7 @@ public Task(
 		ResultPartitionConsumableNotifier resultPartitionConsumableNotifier,
 		PartitionProducerStateChecker partitionProducerStateChecker,
-		Executor executor) {
+		Executor executor) throws IOException {
```
It seems that partition/gate setup causes the exception. Usually, constructors should be able always just construct objects without doing any IO and throwing exceptions. If partition/gate had a lifecycle `setup()` interface method, which would do the same as `NetworkEnviroment.setupPartition/Gate`, then the setup could stay in the same place in `Task.run` as originally, do IO and throw exceptions while running.

zhijiangW replied:
Actually I also do not want to throw execution during constructing `Task`, so in the first version I introduced `Task#initializeInputOutput` for creating and setup partition/gate together. But after reference with partition/gate in task, I reverted the codes as now. I have some concerns for adding `setup` in the interface of partition/gate.

1. I ever thought the lifecycle of partition/gate should be managed inside `ShuffleService`, like `create`, `setup` and `close`, which seems not leaked and self-explanation for different `ShuffleService` implementation. But it would bring overhead for adding `setup` in `ShuffleService` interface.

2. If task could call `setup` of partition/gate directly, it still needs add the partition/gate interface method which might not necessary for other implementations. In addition we need pass more parameters for this `setup`, like `NetworkBufferPool`, `buffersPerGate` etc.

If we add the `setup` lifecycle for task instead of partition/gate, which is similar as my previous `Task#initializeInputOutput`, it might seem easy because we only have one `Task` implementation. In addition, I wish to pass `BufferPool` directly in the construction of partition/gate instead of separate `registerBufferPool` method in future.  What do you think?

azagrebin commented:
@flinkbot approve all

zhijiangW commented:
@azagrebin , I have updated the codes for addressing previous comments.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -89,6 +109,8 @@ public NetworkEnvironment(
 
 		registerNetworkMetrics(metricGroup, networkBufferPool);
 
+		this.ioManager = checkNotNull(ioManager);
```
`ioManager` looks like one more dependency for `ShuffleService` at the moment. Maybe, we could somehow get rid of it in future because it might be mot relevant for all shuffle services. I see it is used in other components. Maybe, we could create a separate `ioManager` inside netty shuffle service.

zhijiangW replied:
Yes, it would be better to create all the required components inside specific `ShuffleService` implementations.

I just quickly reviewed the current usages of `IOManager`. For streaming jobs, the `BarrierBuffer` would be finally not using `IOManager` any more to cache blocked buffers in memory directly. And the `RocksDBStateBackend` only uses the `IOManager` to get path which might be done in other options I think.  So the key blocker for `IOManager` usage is from batch jobs, which relies on it heavily atm. 

Maybe we could create private `IOManager` inside `ShuffleService` as you suggested in future.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -271,4 +378,112 @@ public boolean isShutdown() {
+	/**
+	 * Gauge measuring the number of queued input buffers of a task.
+	 */
+	private static final class InputBuffersGauge implements Gauge<Integer> {
```
I would move these classes to some `network.metrics` package along with `InputChannelMetrics` and `InputGateMetrics`.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskTest.java:
```diff
@@ -251,14 +262,9 @@ public void testExecutionFailsInBlobsMissing() throws Exception {
-		final NetworkEnvironment network = mock(NetworkEnvironment.class);
-		when(network.getResultPartitionManager()).thenReturn(partitionManager);
-		doThrow(new RuntimeException("buffers")).when(network).registerTask(any(Task.class));
+		final NetworkEnvironment network = new NetworkEnvironmentBuilder().build();
```
Only this unit test would shutdown the network atm, so I create a separate network for this created task in order not to affect the shared static `NetworkEnvironment` for other tests.

I also tried to use the shared `NetworkEnvironment` here before and it seems not affect the other tests in practice. I guess the other tests not sensitive with the released `NetworkEnvironment` atm. But for safe consideration in new tests future I created the new network for this special test here. What do you think?

azagrebin replied:
`setup` method should always recreate `networkEnvironment` for each test, so shutdown here should not affect other tests, right? 
btw, `networkEnvironment` does not need to be static and does not seem to be used in any static context.

zhijiangW replied:
Yes, you are right when I double checked the codes. I thought this `NetworkEnvironment` is created only once for all the tests. It should be no problem to create new one for each test. I would fix it now.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskTest.java:
```diff
@@ -940,9 +943,7 @@ private void setState(Task task, ExecutionState state) {
-			final ResultPartitionManager partitionManager = mock(ResultPartitionManager.class);
-			networkEnvironment = mock(NetworkEnvironment.class);
-			when(networkEnvironment.getResultPartitionManager()).thenReturn(partitionManager);
+			network = networkEnvironment;
```
That is fine. 
I could remove `setNetworkEnvironment` in `TaskBuilder` and then make the private variable `network` as final in `TaskBuilder`.  The new constructor `TaskBuilder(NetworkEnvironment)` is defined with the default `newTaskBuilder()` for creating it.

zhijiangW commented:
I have pushed new fixup commit for addressing above comments. 
For the refactor of `TaskBuilder`, I squashed the changes into previous hotfix commit directly.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -367,50 +366,47 @@ public Task(
-
-			++counter;
+		// register detailed network metrics, if configured
+		if (taskManagerConfig.getConfiguration().getBoolean(TaskManagerOptions.NETWORK_DETAILED_METRICS)) {
```
I think it does not hurt always to add only the metric group. I would suggest to move `NETWORK_DETAILED_METRICS` to `NetworkEnvironmentConfiguration`. Then we do not need `optionals` and `NetworkEnvironment.registerIn(Out)putMetrics` can decide whether to add metrics to the group based on `NetworkEnvironmentConfiguration.isDetailedMetrics()`.

zhijiangW replied:
There are two reasons for using `Optional` here:

1. If we pass `Input/OutputGroup` directly, `NETWORK_DETAILED_METRICS` would be checked twice in both `Task` and `NetworkEnvironmentConfiguration`.

2. The `Input/OutputGroup` might be nullable. IMO the optional parameters in interface method might seem better than `Nullable`, also comparing `InputGroup.isPresent()` with `if (InputGroup != null)'.

azagrebin replied:
I meant not to check `NETWORK_DETAILED_METRICS` in Task and always create groups. They will just stay empty if `NetworkEnvironment` adds nothing and it should not hurt. What do you think?

zhijiangW replied:
I misunderstood your previous meaning. 
Hmm it would be better to always create this group in task, then it could keep the same format parameters as other group, also not bring any hurt for users. Totally agree

azagrebin commented:
Thanks for addressing the comments @zhijiangW ! Looks good to me so far!
I suggest we wait for #8181 to be merged and then do final quick check for this PR and merge it.

zhijiangW commented:
Thanks for further reviews and I have rebased the master for merged #8181 @azagrebin

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java:
```diff
@@ -220,6 +222,8 @@ public void setup() throws IOException {
 		haServices.setResourceManagerLeaderRetriever(resourceManagerLeaderRetriever);
 		haServices.setJobMasterLeaderRetriever(jobId, jobManagerLeaderRetriever);
+
+		networkEnvironment = new NetworkEnvironmentBuilder().build();
```
should we call `networkEnvironment.start();`? or at least keep it in `testShouldShutDownTaskManagerServicesInPostStop`? Also in `TaskAsyncCallTest`?

zhijiangW replied:
Actually it is not needed to call start in these tests, because the start would only make `NettyServer` to bind one port to listen and it would not change any internal state. The `NettyServer` is not really needed in these tests. Certainly it would no harm if calling start always. Do you think we should do that in these tests?

zhijiangW replied:
I only call the start for `testShouldShutDownTaskManagerServicesInPostStop` to keep the same behavior as before. 
For the other tests in `TaskTest` and `TaskAsyncCallTest`, I think it is better to not call start atm for the following concerns:
1. Start would bind port which takes some time for each test
2. The bind port would consume more system resources
3. It is not actually useful to start `NettyServer`.

zhijiangW commented:
I updated the codes for addressing the left comments. @azagrebin 

Regarding the `NetworkEnvironment#start` for other tests, I left some concerns. Because of rebasing master to solve the conflicts easily, I squashed all the fixup commits together.
